src/trel_dnssd/trel_dnssd.cpp

```diff
@@ -0,0 +1,542 @@
+                              mRegisterInfo.mTxtEntries);
+}
+
+void TrelDnssd::UnpublishTrelService(void)
+{
+    assert(mRegisterInfo.IsValid());
+    assert(mRegisterInfo.IsPublished());
+
+    mPublisher.UnpublishService(mRegisterInfo.mInstanceName, kTrelServiceName);
+    mRegisterInfo.mInstanceName = "";
+}
+
+void TrelDnssd::OnTrelServiceInstanceAdded(const Mdns::Publisher::DiscoveredInstanceInfo &aInstanceInfo)
+{
+    std::string        instanceName = StringUtils::ToLowercase(aInstanceInfo.mName);
+    otPlatTrelPeerInfo peerInfo;
+
+    // Remove any existing TREL service instance before adding
+    OnTrelServiceInstanceRemoved(instanceName);
+
+    otbrLogDebug("Peer discovered: %s hostname %s addresses %zu port %d priority %d "
+                 "weight %d",
+                 aInstanceInfo.mName.c_str(), aInstanceInfo.mHostName.c_str(), aInstanceInfo.mAddresses.size(),
+                 aInstanceInfo.mPort, aInstanceInfo.mPriority, aInstanceInfo.mWeight);
+
+    for (const auto &addr : aInstanceInfo.mAddresses)
+    {
+        otbrLogDebug("Peer address: %s", addr.ToString().c_str());
+    }
+
+    if (aInstanceInfo.mAddresses.empty())
+    {
+        otbrLogWarning("Peer %s does not have any IPv6 address, ignored", aInstanceInfo.mName.c_str());
+        ExitNow();
+    }
+
+    peerInfo.mRemoved = false;
+    memcpy(&peerInfo.mSockAddr.mAddress, &aInstanceInfo.mAddresses[0], sizeof(peerInfo.mSockAddr.mAddress));
+    peerInfo.mSockAddr.mPort = aInstanceInfo.mPort;
+    peerInfo.mTxtData        = aInstanceInfo.mTxtData.data();
+    peerInfo.mTxtLength      = aInstanceInfo.mTxtData.size();
+
+    {
+        Peer peer(aInstanceInfo.mTxtData, peerInfo.mSockAddr);
+
+        VerifyOrExit(peer.mValid, otbrLogWarning("Peer %s is invalid", aInstanceInfo.mName.c_str()));
+
+        otPlatTrelHandleDiscoveredPeerInfo(mNcp.GetInstance(), &peerInfo);
+
+        mPeers.emplace(instanceName, peer);
+        CheckPeersNumLimit();
+    }
+
+exit:
+    return;
+}
+
+void TrelDnssd::OnTrelServiceInstanceRemoved(const std::string &aInstanceName)
+{
+    std::string instanceName = StringUtils::ToLowercase(aInstanceName);
+    auto        it           = mPeers.find(instanceName);
+
+    VerifyOrExit(it != mPeers.end());
+
+    otbrLogDebug("Peer removed: %s", instanceName.c_str());
+
+    if (CountDuplicatePeers(it->second) == 0)
+    {
+        NotifyRemovePeer(it->second);
+    }
+
+    mPeers.erase(it);
+
+exit:
+    return;
+}
+
+void TrelDnssd::CheckPeersNumLimit(void)
+{
+    const PeerMap::value_type *oldestPeer = nullptr;
+
+    VerifyOrExit(mPeers.size() >= kPeerCacheSize);
+
+    for (const auto &entry : mPeers)
+    {
+        if (oldestPeer == nullptr || entry.second.mDiscoverTime < oldestPeer->second.mDiscoverTime)
+        {
+            oldestPeer = &entry;
+        }
+    }
+
+    OnTrelServiceInstanceRemoved(oldestPeer->first);
```

Comment:
Not possible. `VerifyOrExit(mPeers.size() >= kPeerCacheSize);` already makes sure that `mPeers` is not empty, so that we can always find `oldestPeer`.